### ISSUE TITLE
torch.linalg.qr (float64) repeated in a loop crashes the process (SIGSEGV, exit 139) on ROCm 7.x

### ISSUE THREAD

**Issue by qiyang-ustc** [pytorch/pytorch#185839]
## 🐛 Describe the bug

Calling `torch.linalg.qr` on a float64 matrix repeatedly in a loop causes the **process to be killed by SIGSEGV (exit code 139) after ~31 iterations** on ROCm 7.x (AMD Instinct MI300A). No Python exception is raised — the process dies silently with no traceback.

The crash is **specific to `torch.linalg.qr` in float64**. The following do **not** crash under identical conditions:
- `torch.linalg.qr` in **float32**
- `torch.linalg.svd` / `torch.linalg.eigh` in float64 (also rocSOLVER-backed)
- `torch.matmul` / `@` of any size, with or without `backward()`

It is a **regression from ROCm 6.x**: the same code runs fine on PyTorch 2.6.0+rocm6.2.4 on the same hardware.

## Minimal reproducer

```python
import torch
for i in range(50):
    a = torch.randn(512, 512, device='cuda', dtype=torch.float64)
    torch.linalg.qr(a)
    print(i, flush=True)
```

### Actual output (AMD Instinct MI300A, PyTorch 2.12.0+rocm7.2)

```
0
1
...
29
30
Segmentation fault (core dumped)      # process exits 139, no Python traceback
```

The crash iteration varies slightly by matrix shape (see sweep below) but is reliably reproducible at ~iteration 31 for (512, 512).

## It is not memory exhaustion

`torch.cuda.memory_reserved()` at the crash point is **0.02 GB** (out of 94 GB). The matrices are tiny and freed each iteration.

## Workarounds (both confirmed, 60/60 iterations pass)

```python
# (1) drain pending events periodically
for i in range(50):
    a = torch.randn(512, 512, device='cuda', dtype=torch.float64)
    torch.linalg.qr(a)
    if (i + 1) % 20 == 0:
        torch.cuda.empty_cache()      # <-- prevents the crash

# (2) synchronize each iteration
for i in range(50):
    a = torch.randn(512, 512, device='cuda', dtype=torch.float64)
    torch.linalg.qr(a)
    torch.cuda.synchronize()          # <-- also prevents the crash
```

`PYTORCH_NO_HIP_MEMORY_CACHING=1` also avoids the crash (but is ~5x slower overall).

## The bug is in PyTorch's allocator path, not rocSOLVER itself

A pure C++ `rocsolver_dgeqrf` loop (512×512, float64, `hipMalloc`/`hipFree` each iteration, 60 iterations) runs **cleanly with no crash**. The crash only appears when QR is routed through PyTorch's `CUDACachingAllocator`.

<details>
<summary>C++ rocSOLVER control (passes 60/60)</summary>

```cpp
#include <hip/hip_runtime.h>
#include <rocsolver/rocsolver.h>
// loop 60x: hipMalloc dA,dTau; rocsolver_dgeqrf(handle,512,512,dA,512,dTau);
//           hipDeviceSynchronize(); hipFree(dA); hipFree(dTau);
// Output: "PASSED all 60 iters (pure rocSOLVER C++)"
```
</details>

## Shape dependence

Sweep at 60 iterations, float64 (each shape in its own process):

| shape | result | | shape | result |
|-------|--------|---|-------|--------|
| (512, 512)  | **crash** | | (256, 256)  | pass |
| (500, 500)  | **crash** | | (1024, 64)  | pass |
| (6480, 180) | **crash** | | (2048, 128) | pass |
|             |           | | (1000, 100) | pass |

The crash correlates with the specific rocSOLVER workspace/algorithm path selected for certain (M, N) — consistent with a workspace buffer being tracked on a non-default stream only for those dimensions.

## Root cause analysis

Both `torch.cuda.empty_cache()` and `torch.cuda.synchronize()` eliminate the crash. Both drain the allocator's pending-event queue: `empty_cache()` → `release_cached_blocks()` → `synchronize_and_free_events()`; `synchronize()` completes all GPU work so every queued event is already done.

This points to `c10/cuda/CUDACachingAllocator.cpp`:

1. The float64 QR path allocates a rocSOLVER workspace buffer that is used on a non-default stream, so freeing it calls `insert_events()`, which records a HIP event and pushes `(event, block)` into the `cuda_events` map.
2. On a subsequent allocation, `process_events()` calls `cudaEventQuery(*event)` (→ `hipEventQuery`) on each queued handle to reclaim completed blocks.
3. On **ROCm 6.x**, querying a still-pending event returns `hipErrorNotReady` (handled gracefully). On **ROCm 7.x**, the same query on an accumulated/stale event handle appears to fault inside the HIP runtime (SIGSEGV) before returning an error code — so PyTorch's `else if (err != cudaSuccess)` handler never runs.

float32 QR / SVD / eigh / matmul do not hit this path (different or no `record_stream` on their workspaces), which is why they don't crash.

## Environment

| | |
|---|---|
| PyTorch | 2.12.0+rocm7.2 |
| HIP / ROCm | 7.2.53211 / ROCm 7.2.0 |
| GPU | AMD Instinct MI300A (94 GB) |
| OS | RHEL 8, kernel 4.18.0-553.121.1.el8_10.x86_64 |
| Python | 3.12 |
| Not affected | PyTorch 2.6.0+rocm6.2.4 (same hardware) |

## Things that had no effect

`PYTORCH_TUNABLEOP_ENABLED=0`, `TORCH_BLAS_PREFER_HIPBLASLT=0`,
`PYTORCH_CUDA_ALLOC_CONF=garbage_collection_threshold:{0.5,0.8}`,
`PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:{128,512}`,
`HSA_ENABLE_SDMA=0`, `GPU_MAX_HW_QUEUES=1` — all still crash.

## Possible fixes

- **rocSOLVER/HIP side** (likely the real upstream fix): `hipEventQuery` on a pending event should return `hipErrorNotReady`, matching CUDA and ROCm 6.x. The process-fatal behavior is itself a regression worth filing against ROCm.
- **PyTorch side** (mitigation): make `process_events()` more robust on ROCm — e.g. gate the query behind `hipStreamQuery` for the recording stream, or drain `cuda_events` via `synchronize_and_free_events()` once the queue exceeds a small threshold under `#if defined(USE_ROCM)`.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @jianyuh @nikitaved @mruberry @walterddr @Lezcano

**Comment by wasimat404** [comment]
Edit: Retracting my earlier analysis. A gdb backtrace on real MI300A hardware (shared by the reporter) shows the fault is in amd::roc::VirtualGPU::HwQueueTracker::ActiveSignal during orgqr's async kernel dispatch — not in process_events() / hipEventQuery as I suggested. The allocator mitigation I proposed would not address the actual crash. The root cause appears to be a race in the HIP runtime's signal/queue tracking, triggered by the high-rate GEMM launches that rocsolver_orgqr produces. This likely belongs against ROCm/clr or rocSOLVER, not PyTorch.

**Comment by qiyang-ustc** [comment]
> Edit: Retracting my earlier analysis. A gdb backtrace on real MI300A hardware (shared by the reporter) shows the fault is in amd::roc::VirtualGPU::HwQueueTracker::ActiveSignal during orgqr's async kernel dispatch — not in process_events() / hipEventQuery as I suggested. The allocator mitigation I proposed would not address the actual crash. The root cause appears to be a race in the HIP runtime's signal/queue tracking, triggered by the high-rate GEMM launches that rocsolver_orgqr produces. This likely belongs against ROCm/clr or rocSOLVER, not PyTorch.

Thank you sincerely for your analysis! I think you are right that this issue should be submitted to ROCM, but a temporal fixing here might be useful for other users. So I think I can left this issue open here?

**Comment by wasimat404** [comment]
Agreed on leaving it open as a tracking issue  a workaround note here helps others hitting this until the runtime fix lands. For anyone blocked right now, the reliable mitigations are AMD_SERIALIZE_KERNEL=3, HIP_LAUNCH_BLOCKING=1, or a torch.cuda.synchronize() after the QR call , all of which serialize the dispatch that triggers the HwQueueTracker::ActiveSignal race.
The actual fix likely belongs in the HIP runtime (ROCm/clr) rather than PyTorch, since forcing synchronization inside the orgqr path would mask the race at the cost of QR throughput. @jeffdaily @jithunnair-amd would know whether this is already tracked on the ROCm side.

**Comment by naromero77amd** [comment]
@wasimat404 @qiyang-ustc If either of you have filed a GH issue with the other ROCm projects, please link them here. Thanks.

**Comment by naromero77amd** [comment]
Able to reproduce this issue on an MI200. Looks like rocprofiler-sdk is the problem. Able to work around the issue by taking the reproducer and setting `HSA_TOOLS_DISABLE_REGISTER=1`.

Same issue as: https://github.com/ROCm/rocm-systems/issues/5071

**Comment by naromero77amd** [comment]
@qiyang-ustc This issue is fixed in nightly ROCm. You can varifiy this for yourself by downloading this docker image:
```
ghcr.io/pytorch/ci-image:pytorch-linux-noble-rocm-nightly-py3-67c24048dab0d5f643710517275cc9a6810d98b1
```

Then building PyTorch against it. I tried this with PyTorch 2.11 and can confirm that the problem is fixed.

We can keep this open until this is available in the upsteam wheels.

**Comment by mwootton** [comment]
I re-engineered the fix from the reproducer.  It is exactly this:
https://github.com/ROCm/rocm-systems/pull/4066
